### PR TITLE
Update Projects According to Piazza Discussions

### PR DESCRIPTION
**Pull request by xinzhu-cai**
Main changes:
1. Update comments to clarify data structures and function implementation details 
2. Added several cpp files for executors (Project 3) so that students can work on cpp files only

### CODE REVIEW

**Comment by xinzhu-cai** [comment]
Added a binary data copying test to Proj1
Implemented buffer pool manager callbacks for grading usage

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.10)
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)   # For clang-tidy.
-
+set(BUILD_SHARED_LIBS ON)
```
Comment (apavlo) [inline CMakeLists.txt]: Why was this added?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -11,3 +11,4 @@ file(GLOB_RECURSE murmur3_sources
         ${PROJECT_SOURCE_DIR}/third_party/murmur3/*.cpp ${PROJECT_SOURCE_DIR}/third_party/murmur3/*.h)
 add_library(thirdparty_murmur3 SHARED ${murmur3_sources})
 target_link_libraries(bustub_shared thirdparty_murmur3)
+target_link_libraries(bustub_shared gtest)
```
Comment (apavlo) [inline src/CMakeLists.txt]: Remove this. There should **not** be any testing code in the `src` directory.

File: src/buffer/buffer_pool_manager.cpp [hunk 3]
```diff
@@ -42,13 +42,22 @@ Page *BufferPoolManager::FetchPageImpl(page_id_t page_id) {
   // 2.     If R is dirty, write it back to the disk.
   // 3.     Delete R from the page table and insert P.
   // 4.     Update P's metadata, read in the page content from disk, and then return a pointer to P.
+
+  // For grading. Do not Modify!
+  counter.AddCount(FuncType::FetchPage);
```
Comment (apavlo) [inline src/buffer/buffer_pool_manager.cpp]: This should **not** be in the implementation method (`FetchPageImpl`). This should be in the public API method (`FetchPage`).

File: src/include/buffer/buffer_pool_manager.h [hunk 4]
```diff
@@ -17,20 +17,117 @@
...
+      }
+    }
+
+    void CheckFlushAllPages() {
+      for (int i = 1; i < 5; ++i) {
+        EXPECT_EQ(counts[i], 0) << "FlushAllPage Should not call other functions";
+      }
+      EXPECT_EQ(counts[5], 1) << "has to call FlushAllPage once";
+    }
+  };
```
Comment (apavlo) [inline src/include/buffer/buffer_pool_manager.h]: There should be **no** testing code inside of the `src` directory. Remove the dependency on "gtest".

File: src/include/buffer/buffer_pool_manager.h [hunk 5]
```diff
@@ -17,20 +17,117 @@
...
+          break;
+        case FuncType::DeletePage:
+          counter.CheckDeletePage();
+          break;
+        case FuncType::FlushAllPages:
+          counter.CheckFlushAllPages();
+          break;
+      }
+    }
+  }
```
Comment (apavlo) [inline src/include/buffer/buffer_pool_manager.h]: Move the counter callback to be outside of the class. Define it in a separate file. It is also only used for testing, so we can potentially move into the `test` directory.

We want to reduce the amount of code in the parts of the system that the students have to look at to minimize the complexity.

File: src/include/buffer/buffer_pool_manager.h [hunk 6]
```diff
@@ -44,51 +141,52 @@ class BufferPoolManager {
    */
   ~BufferPoolManager();
 
-  /** Grading function. Do not modify! */
-  Page *FetchPage(page_id_t page_id, bufferpool_callback_fn callback = nullptr) {
-    GradingCallback(callback, CallbackType::BEFORE, page_id);
+  /** Grading function. Do not modify/call! */
+  Page *FetchPage(page_id_t page_id, bufferpool_callback_fn callback = &BufferPoolManager::counter_callback) {
```
Comment (apavlo) [inline src/include/buffer/buffer_pool_manager.h]: The default callback should probably be `nullptr`.

File: src/include/buffer/buffer_pool_manager.h [hunk 7]
```diff
@@ -17,20 +17,117 @@
...
 /**
  * BufferPoolManager reads disk pages to and from its internal buffer pool.
  */
 class BufferPoolManager {
+ private:
+  struct Counter {
+    // 0-FetchPage  1-UnpinPage  2-FlushPage 3-NewPage 4-DeletePage
+    // 5-FlushAllPages
+    static const int num_types = 6;
+    int counts[num_types];
```
Comment (apavlo) [inline src/include/buffer/buffer_pool_manager.h]: Don't use implicit values to define specific things. Use `enum class`

File: src/include/buffer/buffer_pool_manager.h [hunk 8]
```diff
@@ -17,20 +17,117 @@
...
+      EXPECT_EQ(counts[5], 1) << "has to call FlushAllPage once";
+    }
+  };
+
  public:
   enum class CallbackType { BEFORE, AFTER };
-  using bufferpool_callback_fn = void (*)(enum CallbackType, const page_id_t page_id);
+  using bufferpool_callback_fn = void (BufferPoolManager::*)(enum CallbackType type, FuncType func_type);
 
+  void counter_callback(enum CallbackType type, FuncType func_type) {
```
Comment (apavlo) [inline src/include/buffer/buffer_pool_manager.h]: There may be other callbacks that we want to execute other than just counting invocations. For example, we may have a test case that inspects internal data structures using callbacks.

File: src/include/buffer/buffer_pool_manager.h [hunk 9]
```diff
@@ -17,20 +17,117 @@
...
 
 #include "buffer/clock_replacer.h"
+#include "gtest/gtest.h"
 #include "recovery/log_manager.h"
 #include "storage/disk/disk_manager.h"
 #include "storage/page/page.h"
 
 namespace bustub {
 
+enum FuncType { FetchPage, UnpinPage, FlushPage, NewPage, DeletePage, FlushAllPages };
```
Comment (apavlo) [inline src/include/buffer/buffer_pool_manager.h]: Switch to `enum class`.
Also, `FuncType` is a generic name but these functions are specific to the `BufferPoolManager`. You should rename to `BufferPoolManagerMethodType`.

File: src/include/execution/executors/seq_scan_executor.h [hunk 10]
```diff
@@ -31,13 +31,13 @@ class SeqScanExecutor : public AbstractExecutor {
...
-  SeqScanExecutor(ExecutorContext *exec_ctx, const SeqScanPlanNode *plan) : AbstractExecutor(exec_ctx) {}
+  SeqScanExecutor(ExecutorContext *exec_ctx, const SeqScanPlanNode *plan);
 
-  void Init() override {}
+  void Init() override;
 
-  bool Next(Tuple *tuple) override { return false; }
+  bool Next(Tuple *tuple) override;
 
-  const Schema *GetOutputSchema() override { return plan_->OutputSchema(); }
```
Comment (apavlo) [inline src/include/execution/executors/seq_scan_executor.h]: It is okay to leave this method in the header file because it's not something the students will have to implement.

File: src/include/recovery/log_record.h [hunk 11]
```diff
@@ -99,14 +99,14 @@ class LogRecord {
...
+  LogRecord(txn_id_t txn_id, lsn_t prev_lsn, LogRecordType log_record_type, page_id_t prev_page_id)
       : size_(HEADER_SIZE),
         txn_id_(txn_id),
         prev_lsn_(prev_lsn),
         log_record_type_(log_record_type),
-        prev_page_id_(page_id) {
+        prev_page_id_(prev_page_id) {
     // calculate log record size
-    size_ = HEADER_SIZE + sizeof(page_id_t);
+    size_ = HEADER_SIZE + sizeof(page_id_t) * 2;
```
Comment (apavlo) [inline src/include/recovery/log_record.h]: Add a comment that explains why `* 2`

File: src/include/storage/page/hash_table_page_defs.h [hunk 12]
```diff
@@ -14,6 +14,8 @@
 
 #define MappingType std::pair<KeyType, ValueType>
 
+/** BLOCK_ARRAY_SIZE is the number of (key, value) pairs that can be stored in a block page. It is an approximate calculation based on the size of MappingType (which is a std::pair of KeyType and ValueType). For each key/value pair, we need two additional bits for occupied_ and readable_.
```
Comment (apavlo) [inline src/include/storage/page/hash_table_page_defs.h]: Fix formatting of this comment.

File: src/recovery/log_manager.cpp [hunk 13]
```diff
@@ -16,9 +16,9 @@ namespace bustub {
 /*
  * set enable_logging = true
  * Start a separate thread to execute flush to disk operation periodically
- * The flush can be triggered when the log buffer is full or buffer pool
- * manager wants to force flush (it only happens when the flushed page has a
- * larger LSN than persistent LSN)
+ * The flush can be triggered when timeout or the log buffer is full or buffer * pool manager wants to force flush (it only happens when the flushed page has * a larger LSN than persistent LSN)
```
Comment (apavlo) [inline src/recovery/log_manager.cpp]: Fix formatting of this comment.

File: test/buffer/buffer_pool_manager_test.cpp [hunk 14]
```diff
@@ -14,10 +14,69 @@
...
+  ASSERT_NE(nullptr, page0);
+  EXPECT_EQ(0, page_id_temp);
+
+  char random_binary_data[PAGE_SIZE];
+  unsigned int seed = 15645;
+  for (char &i : random_binary_data) {
+    i = static_cast<char>(rand_r(&seed) % 256);
+  }
+
+  random_binary_data[PAGE_SIZE / 2] = '\0';
```
Comment (apavlo) [inline test/buffer/buffer_pool_manager_test.cpp]: Add comment that explains what you are doing here.

File: test/buffer/buffer_pool_manager_test.cpp [hunk 15]
```diff
@@ -14,10 +14,69 @@
...
+  auto *bpm = new BufferPoolManager(buffer_pool_size, disk_manager);
+
+  page_id_t page_id_temp;
+  auto *page0 = bpm->NewPage(&page_id_temp);
+
+  // Scenario: The buffer pool is empty. We should be able to create a new page.
+  ASSERT_NE(nullptr, page0);
+  EXPECT_EQ(0, page_id_temp);
+
+  char random_binary_data[PAGE_SIZE];
```
Comment (apavlo) [inline test/buffer/buffer_pool_manager_test.cpp]: Add comment that explains what you are doing.

File: src/include/execution/executors/insert_executor.h [hunk 16]
```diff
@@ -34,16 +34,15 @@ class InsertExecutor : public AbstractExecutor {
    * @param child_executor the child executor to obtain insert values from, can be nullptr
    */
   InsertExecutor(ExecutorContext *exec_ctx, const InsertPlanNode *plan,
-                 std::unique_ptr<AbstractExecutor> &&child_executor)
-      : AbstractExecutor(exec_ctx) {}
+                 std::unique_ptr<AbstractExecutor> &&child_executor);
 
-  const Schema *GetOutputSchema() override { return plan_->OutputSchema(); }
```
Comment (apavlo) [inline src/include/execution/executors/insert_executor.h]: It is okay to leave this method in the header file because it's not something the students will have to implement.

File: src/include/execution/executors/hash_join_executor.h [hunk 17]
```diff
@@ -96,17 +96,16 @@ class HashJoinExecutor : public AbstractExecutor {
...
    */
   HashJoinExecutor(ExecutorContext *exec_ctx, const HashJoinPlanNode *plan, std::unique_ptr<AbstractExecutor> &&left,
-                   std::unique_ptr<AbstractExecutor> &&right)
-      : AbstractExecutor(exec_ctx) {}
+                   std::unique_ptr<AbstractExecutor> &&right);
 
   /** @return the JHT in use. Do not modify this function, otherwise you will get a zero. */
   // Uncomment me! const HT *GetJHT() const { return &jht_; }
 
-  const Schema *GetOutputSchema() override { return plan_->OutputSchema(); }
```
Comment (apavlo) [inline src/include/execution/executors/hash_join_executor.h]: It is okay to leave this method in the header file because it's not something the students will have to implement.

File: src/include/execution/executors/aggregation_executor.h [hunk 18]
```diff
@@ -161,35 +161,22 @@ class AggregationExecutor : public AbstractExecutor {
...
+  const Schema *GetOutputSchema() override;
 
-  void Init() override {}
+  void Init() override;
 
-  bool Next(Tuple *tuple) override { return false; }
+  bool Next(Tuple *tuple) override;
 
   /** @return the tuple as an AggregateKey */
-  AggregateKey MakeKey(const Tuple *tuple) {
```
Comment (apavlo) [inline src/include/execution/executors/aggregation_executor.h]: It is okay to leave this method in the header file because it's not something the students will have to implement.

File: src/include/execution/executors/aggregation_executor.h [hunk 19]
```diff
@@ -161,35 +161,22 @@ class AggregationExecutor : public AbstractExecutor {
...
-    std::vector<Value> keys;
-    for (const auto &expr : plan_->GetGroupBys()) {
-      keys.emplace_back(expr->Evaluate(tuple, child_->GetOutputSchema()));
-    }
-    return {keys};
-  }
+  AggregateKey MakeKey(const Tuple *tuple);
 
   /** @return the tuple as an AggregateValue */
-  AggregateValue MakeVal(const Tuple *tuple) {
```
Comment (apavlo) [inline src/include/execution/executors/aggregation_executor.h]: It is okay to leave this method in the header file because it's not something the students will have to implement.

File: src/include/buffer/buffer_pool_manager.h [hunk 20]
```diff
@@ -152,6 +251,8 @@ class BufferPoolManager {
    */
   void FlushAllPagesImpl();
 
+  // For grading. Do not modify!
+  Counter counter;
```
Comment (apavlo) [inline src/include/buffer/buffer_pool_manager.h]: Remove this.

File: src/buffer/buffer_pool_manager.cpp [hunk 21]
```diff
@@ -42,13 +42,22 @@ Page *BufferPoolManager::FetchPageImpl(page_id_t page_id) {
...
+bool BufferPoolManager::UnpinPageImpl(page_id_t page_id, bool is_dirty) {
+  // For grading. Do not Modify!
+  counter.AddCount(FuncType::UnpinPage);
+  return false;
+}
 
 bool BufferPoolManager::FlushPageImpl(page_id_t page_id) {
   // Make sure you call DiskManager::WritePage!
+  // For grading. Do not Modify!
+  counter.AddCount(FuncType::FlushPage);
```
Comment (apavlo) [inline src/buffer/buffer_pool_manager.cpp]: Remove.

File: src/buffer/buffer_pool_manager.cpp [hunk 22]
```diff
@@ -58,6 +67,9 @@ Page *BufferPoolManager::NewPageImpl(page_id_t *page_id) {
   // 2.   Pick a victim page P from either the free list or the replacer. Always pick from the free list first.
   // 3.   Update P's metadata, zero out memory and add P to the page table.
   // 4.   Set the page ID output parameter. Return a pointer to P.
+
+  // For grading. Do not Modify!
+  counter.AddCount(FuncType::NewPage);
```
Comment (apavlo) [inline src/buffer/buffer_pool_manager.cpp]: Remove.

File: src/buffer/buffer_pool_manager.cpp [hunk 23]
```diff
@@ -67,11 +79,16 @@ bool BufferPoolManager::DeletePageImpl(page_id_t page_id) {
   // 1.   If P does not exist, return true.
   // 2.   If P exists, but has a non-zero pin-count, return false. Someone is using the page.
   // 3.   Otherwise, P can be deleted. Remove P from the page table, reset its metadata and return it to the free list.
+
+  // For grading. Do not Modify!
+  counter.AddCount(FuncType::DeletePage);
```
Comment (apavlo) [inline src/buffer/buffer_pool_manager.cpp]: Remove.

File: src/buffer/buffer_pool_manager.cpp [hunk 24]
```diff
@@ -67,11 +79,16 @@ bool BufferPoolManager::DeletePageImpl(page_id_t page_id) {
...
   return false;
 }
 
 void BufferPoolManager::FlushAllPagesImpl() {
   // You can do it!
+  // For grading. Do not Modify!
+  counter.AddCount(FuncType::FlushAllPages);
```
Comment (apavlo) [inline src/buffer/buffer_pool_manager.cpp]: Remove.

File: CMakeLists.txt (hunk 1 above)
Comment (xinzhu-cai) [reply CMakeLists.txt]: To prevent "Linking CXX shared library error; recompile with -fPIC"

**Comment by apavlo** [comment]
@xinzhu-cai Can you mark the comments from my previous review as 'Resolved'?
